pkg/controller/cronjob/utils.go

```diff
@@ -148,6 +148,33 @@ func getRecentUnmetScheduleTimes(sj batchv1beta1.CronJob, now time.Time) ([]time
 	return starts, nil
 }
 
+// LatestScheduleTime, foundSchedule
+// getLatestMissedSchedule returns the latest start time in the specified time window (inclusive), and a true/false indicating is a start time was found.
+// It uses a binary search with the cron.Next function, to avoid stepping through every cron schedule.
+// This is less efficient than if cron.Previous existed, but is arguable simpler to test than a custom Previous function.
+func getLatestMissedSchedule(startWindow time.Time, endWindow time.Time, schedule cron.Schedule) (time.Time, bool) {
```

Comment:
We should probably return (time.Time, int) and the int should be the number of possible start times we found. It won't be the exact total, it'll just be the ones we found.

Reply:
And we should probably include a merge function that takes the two and returns the correct merge, something like: `func merge(t1, t2 time.Time, c1, c2 int) (time.Time, int)` which returns the earliest of the t's and sums the c's

Reply:
The goal is to get the latest time, since that's the time that the controller is acting "on behalf of" (for naming & "last scheduled" status).

Are you thinking about using the count for the events/logs? We care about 0, 1, multiple... this has an edge case where the count would be 1. But we could potentially fold that process in to reduce the Next calls, EG only "double check" if count=1 and the first half was never checked.

Reply:
> Are you thinking about using the count for the events/logs? We care about 0, 1, multiple... this has an edge case where the count would be 1. But we could potentially fold that process in to reduce the Next calls, EG only "double check" if count=1 and the first half was never checked.

Yes exactly. I think we need to know the number of potential calls so that the logging is the same ("missed multiple start times")

Reply:
Not perfect, but something like this? https://gist.github.com/tomwans/604b0661747363af6438042a87dcd6a8

Reply:
I implemented something along these lines - call Next() once, then go into the binary search from $result-onwards.